File: src/textSimilarity/similarities.ts
import { getIntersection, getUnion, dotProduct, magnitude } from "./utils";

export function jaccardSimilarity(tokens1: string[], tokens2: string[]) {
    const intersection = getIntersection(tokens1, tokens2);
    const union = getUnion(tokens1, tokens2);

    return intersection.size / union.size;
}

// Más permisiva que Jaccard (buena para frases cortas)
export function diceCoefficient(tokens1: string[], tokens2: string[]) {
    const intersection = getIntersection(tokens1, tokens2);

    return intersection.size / (tokens1.length + tokens2.length);
}

// Buena para frases muy cortas
export function overlapCoefficient(tokens1: string[], tokens2: string[]) {
    const intersection = getIntersection(tokens1, tokens2);

    const min = Math.min(tokens1.length, tokens2.length);
    return intersection.size / min;
}

// Mide dirección, pero ignora magnitud
export function cosineSimilarity(vector1: number[], vector2: number[]) {
    if (vector1.length !== vector2.length) {
        throw new Error("Vectors must have same length.");
    }
    return dotProduct(vector1, vector2) / (magnitude(vector1) * magnitude(vector2));
}

// Mide distancia real
export function euclideanSimilarity(vector1: number[], vector2: number[]) {
    if (vector1.length !== vector2.length) {
        throw new Error("Vectors must have same length.");
    }

    let sumSquares = 0;
    for (let i = 0; i < vector1.length; ++i) {
        sumSquares += (vector1[i] - vector2[i]) ** 2;
    }
    const distance = Math.sqrt(sumSquares);
    // Convertir distancia a similitud [0,1]
    return 1 / (1 + distance);
}

// Mide distancia real
export function manhattanSimilarity(vector1: number[], vector2: number[]) {
    if (vector1.length !== vector2.length) {
        throw new Error("Vectors must have same length.");
    }

    let distance = 0;
    for (let i = 0; i < vector1.length; ++i) {
        distance += Math.abs(vector1[i] - vector2[i]);
    }
    // Convertir distancia a similitud [0,1]
    return 1 / (1 + distance);
}

export function dotProductSimilarity(vector1: number[], vector2: number[]) {
    if (vector1.length !== vector2.length) {
        throw new Error("Vectors must have same length.");
    }
    // Devolver el producto escalar (magnitud importa).
    // No está normalizada a [0,1]
    return dotProduct(vector1, vector2);
}

File: server/similarities.py
from math_utils import get_intersection, get_union, euclidean_distance, manhattan_distance
import numpy as np
import numpy.typing as npt

# --- Token-based similarities ---

def jaccard_similarity(tokens1: list[str], tokens2: list[str]):
    intersection = get_intersection(tokens1, tokens2)
    union = get_union(tokens1, tokens1)

    return len(intersection) / len(union)

def dice_coefficient(tokens1: list[str], tokens2: list[str]):
    intersection = get_intersection(tokens1, tokens2)

    return 2 * len(intersection) / (len(tokens1) + len(tokens2))

def overlap_coefficient(tokens1: list[str], tokens2: list[str]):
    intersection = get_intersection(tokens1, tokens2)
    min_len = min(len(tokens1), len(tokens2))

    return len(intersection) / min_len

# --- Vector-based similarities ---

def cosise_similarity(vector1: list[float], vector2: list[float]):
    array1 = np.array(vector1)
    array2 = np.array(vector2)
    if array1.shape != array2.shape:
        raise ValueError("Vectors must have the same length.")
    dot = np.dot(vector1, array2)
    distance = euclidean_distance(array1 - array2)
    return dot / distance

def euclidean_similarity(vector1: list[float], vector2: list[float]):
    array1 = np.array(vector1)
    array2 = np.array(vector2)
    if array1.shape != array2.shape:
        raise ValueError("Vectors must have the same length.")
    distance = euclidean_distance(array1 - array2)
    return 1 / (1 + distance)

def manhattan_similarity(vector1: list[float], vector2: list[float]):
    array1 = np.array(vector1)
    array2 = np.array(vector2)
    if array1.shape != array2.shape:
        raise ValueError("Vectors must have the same length.")
    distance = manhattan_distance(array1 - array2)
    return 1 / (1 + distance)